### DISCUSSION TITLE
Support of base URL does not work as expected.

### DISCUSSION THREAD

**Discussion by Tchoupinax** [gethomepage/homepage#3102] [Support]
### Description

Hello,

I try to setup homepage behind a subpath and for that I need to configure a base URL. I found into the documentation and Github issues that there are two options about that: `base` and `startURL`. However, any of thoses impact my configuration.

I found [this discussion ](https://github.com/gethomepage/homepage/discussions/150) but the solution does not change.

I can see that the attribute `base` is correctly changed in my HTML page but resources URL are not impacted.

Also, I tried to edit the next configuration and build a new customized image:
- Before: Resources are not loaded (css, js) => the prefix misses
- After: Resources are loaded so the page works but api is not reachable, the prefix misses for api here

```bash
const nextConfig = {
   reactStrictMode: true,
   output: "standalone",
+  basePath: "/my-custom-path",
   images: {
     domains: ["cdn.jsdelivr.net"],
     unoptimized: true,
   },
   i18n,
};
```

When I add the base url, it's like this
![image](https://github.com/gethomepage/homepage/assets/9136206/e2a99679-e16c-4884-80af-25c42473872c)
and the result is here
![image](https://github.com/gethomepage/homepage/assets/9136206/f5a253eb-f658-4272-bbea-5d691a9444dc)

### homepage version

v0.8.9

### Installation method

Docker

### Configuration

```yaml
settings.yaml


---
# For configuration options and examples, please see:
# https://gethomepage.dev/latest/configs/settings

title: My custom dashbaord
base: http://localhostd:3000/totot

disableCollapse: true
fiveColumns: true
headerStyle: clean
statusStyle: "dot"
useEqualHeights: true
```

### Troubleshooting

All API requests returns a 404 and it's normal because the frontend should prefix the request by the prefix (included in `base` or `startUrl`.

That's said, I reached out this state because I build a **customized version** with the prefix in the next config

![image](https://github.com/gethomepage/homepage/assets/9136206/1f085cff-aff5-4d4f-8739-8ec2572d6101)

**Comment by glime-ai** [comment]
Thanks for posting a discussion about homepage here on GitHub. Homepage now uses a fancy new AI tool thanks to our friends at Glime, so the answer below is the AI bot's attempt to answer your question. If you feel it's sufficient you can mark the answer and close this post, otherwise feel free to leave it open of course!

Thanks for using Homepage! ❤️

---

Based on the context provided, it seems you are trying to set up the homepage application to work behind a subpath using Docker and are encountering issues with the `base` and `startUrl` settings not affecting your configuration as expected.

The `base` setting is indeed intended to change the base URL of your homepage instance, and the `startUrl` is used to customize the start URL for installable apps. However, it appears that setting these options in your `settings.yaml` file is not having the desired effect on resource URLs or API endpoints.

From the information provided, it seems you have attempted to modify the Next.js configuration directly by adding a `basePath` to the `nextConfig` object. This is a correct approach when you need to serve a Next.js application from a subpath. However, it's important to note that this change alone might not be sufficient if the application's server-side logic does not account for the new base path.

The images you've linked to show that the resources are not being loaded with the correct base path, which suggests that the application's static assets and API requests are not being properly prefixed.

To resolve this issue, you should ensure that:

1. The `base` setting in your `settings.yaml` file is correctly set to the subpath you want to use (e.g., `/my-custom-path`).
2. The `basePath` in the Next.js configuration is also set to the same subpath.
3. The application's server-side logic (including API endpoints) respects the `basePath` and correctly prefixes all routes with it.

If the application's server-side logic does not automatically handle the `basePath`, you may need to manually update the API endpoints and any other server-side routes to include the subpath.

Since you are building a customized version of the application, you may also need to review the server-side code to ensure that it is aware of the `basePath` and correctly constructs URLs for static assets and API requests.

If you continue to experience issues, you may want to reach out to the homepage community for support or consult the documentation for more detailed guidance on configuring the application for subpath deployment.

Unfortunately, without access to the actual repository and codebase, I cannot provide a more specific solution. Please ensure that you are following the latest documentation and consider opening an issue in the homepage repository if you believe there is a bug or if the documentation is lacking in this area.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=Bl3mlvfCnTnRm5&utm_medium=github).

**Reply by CosasDePuma** [reply]
Any update? This tips don't solve the issue.

**Comment by omidforoqi** [comment]
+1

**Comment by TomasMonkevic** [comment]
Having the same issue with traefik

**Comment by karco** [comment]
I'm very happy to find people with the same problem as my installation.   I've spent a lot of time on it, but I still don't have a solution to display the page under .../homepage.

`- “traefik.http.routers.homepage.rule=(Host(`$HOSTNAME.$DOMAINNAME_1`) && PathPrefix(`/homepage`)) || Host(`homepage.$DOMAINNAME_1`)”`

**Comment by fussel132** [comment]
Same here, no matter which option is used (base with(out) startUrl), this is not working with the ghcr.io image

**Comment by fussel132** [comment]
@benphelps you introduced the feature in #150, I saw that in the meantime there was a fix by @shamoon in #2194 which closed issues #2128 and #539, both related to the base url not properly. It might work with nginx now but unfortunately not with traefik.

When I set the base url (to like domain.com/homepage) there is no base tag visible in my HTML. Using the traefik labels from #150 I see the default homepage with no config applied and just the html (it cannot find the css obviously) until I open the homepage once via the forwarded docker port directly and not through traefik. Only then I can see the custom config applied for the proxied homepage (which is still missing everything else than the base html). Strange.

And I shouldn't need to compile from source as #2194 changed the base url in next.config.js to a relative path right?

**Comment by fussel132** [comment]
In case someone still struggles to get this to work:
As a temporary workaround I've replaced all the hard-coded /api/ paths in the code with /admin/api/ as well as setting the next.js basePath to /admin so this now works on desired basePath. See commits [here](https://github.com/fussel132/homepage-baseurl) if you want to use this/modify this on your own.

**Comment by BenLikeCode** [comment]
same problem. 
My conf:

DNS -> reverse proxy traefik (physical server)-> Nodeport traefik (minikube) -> Homepage service (minikube)

Do you know an alternative that works easily with traefik?

**Comment by juntong20XX** [comment]
Hi all,

I'm trying to install homepage at `/home/`, based on #539 , now the following traefik config works for me:

```
- "traefik.http.routers.home-proxy.entrypoints=web"
- "traefik.http.routers.home-proxy.rule=Host(`example.com`) && (PathPrefix(`/home`) || PathPrefix(`/api`) || PathPrefix(`/_next`))"
- "traefik.http.routers.home-proxy.middlewares=gzip@file, home-skip"
- "traefik.http.routers.home-proxy.service=home"
# Middlewares
- "traefik.http.middlewares.home-skip.stripprefix.prefixes=/home"
```

with `base: https://example.com/home/` in settings.yaml

Holp that can help you. 😃

**Comment by FL550** [comment]
I can also provide this working docker-compose config for traefik working on custom port rather than a subdomain :

```yaml
  homepage:
    image: ghcr.io/gethomepage/homepage:latest
    container_name: homepage
    ports:
      - 8080:3000
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock # pass local proxy
    restart: always
    environment:
      HOMEPAGE_ALLOWED_HOSTS: 192.168.1.X:8080
    labels:
      - traefik.enable=true
      - traefik.http.routers.homepage.rule=PathPrefix(`/api`) || PathPrefix(`/_next`))
      - traefik.http.routers.homepage.service=homepage  
```

with `base: http://192.168.1.X:8080/` set in `settings.yaml`.

**Comment by xXxNIKIxXx** [comment]
Hey any update of an implimentation using env vars?

**Comment by ChilledSlim** [comment]
The problem with these "solutions" is that they still use /api and /_next at the root path.
So it will still conflict with things it shouldn't and doesn't isolate homepage under a path prefix as people want/need.

**Comment by shamoon** [comment]
The AI answer above is still accurate. AFAIK there is no way to do this without re-building the next-js frontend. If someone has another solution that would be welcome, but I believe there simply isnt one

**Comment by ileodo** [comment]
I know this is a such hack, but if you are using nginx as reverse proxy, you can try this magic:

```
    proxy_set_header Accept-Encoding "";
    sub_filter_once off;
    sub_filter_types *;

    sub_filter '"src":"/' '"src":"/homepage/';
    sub_filter 'href:"/' 'href:"/homepage/';
    sub_filter 'href="/' 'href="/homepage/';
    sub_filter '"/api/' '"/homepage/api/';
    sub_filter '`/api/' '`/homepage/api/';
    sub_filter '"/_next/' '"/homepage/_next/';
    sub_filter 'url(/_next/' 'url(/homepage/_next/';
    sub_filter '"/icons/' '"/homepage/icons/';
    sub_filter '"/images/' '"/homepage/images/';
    sub_filter '"static/chunks/"' '"../homepage/_next/static/chunks/"';
```

**Reply by pascal-cm** [reply]
> I know this is a such hack, but if you are using nginx as reverse proxy, you can try this magic:
> 
> ```
>     proxy_set_header Accept-Encoding "";
>     sub_filter_once off;
>     sub_filter_types *;
> 
>     sub_filter '"src":"/' '"src":"/homepage/';
>     sub_filter 'href:"/' 'href:"/homepage/';
>     sub_filter 'href="/' 'href="/homepage/';
>     sub_filter '"/api/' '"/homepage/api/';
>     sub_filter '`/api/' '`/homepage/api/';
>     sub_filter '"/_next/' '"/homepage/_next/';
>     sub_filter 'url(/_next/' 'url(/homepage/_next/';
>     sub_filter '"/icons/' '"/homepage/icons/';
>     sub_filter '"/images/' '"/homepage/images/';
>     sub_filter '"static/chunks/"' '"../homepage/_next/static/chunks/"';
> ```

FWIW, I believe my feature request (working fork included) resolves this issue as it lets you deploy homepage under any path. No more filtering or mapping URLs: https://github.com/gethomepage/homepage/discussions/6625
